### ISSUE TITLE
Proposal for a UPS overhaul

### ISSUE THREAD

**Issue by Inglonias** [CleverRaven/Cataclysm-DDA#31339]
**Is your feature request related to a problem? Please describe.**  
Ever since the battery overhaul, the UPS has been a bit of a problem in terms of how or whether it works. In lore, it's supposed to be the battery equivalent of a USB, a central source for power for devices that use it. Right now, it's just some weird thing that kinda sorta does stuff if you squint. Maybe.

**Describe the solution you'd like**  
The UPS needs a purpose again. My proposal is that we turn it into a power bank, in the sense that you can deposit and withdraw electric charge from it at will from any other batteries you possess. It would have a very large capacity, but cannot be utilized directly - You would need devices designed or modified to use it.

A UPS is essentially a device that can directly store, collect, and distribute electric charge. It would have a high capacity. Very high. High enough to power several tools at once for a while if you want.

You may activate the UPS and from there, select "Deposit" or "Withdraw". If you select "Deposit", you are given a list of all batteries in your inventory that contain any amount of charge. Selecting one lets you drain that battery into the UPS. If you select "Withdraw", you are again given a list of all batteries in your inventory, but this time, the list is of batteries that are not full. Selecting a battery will let you drain the UPS into that battery. This makes the UPS useful for consolidating and moving electric charge quickly and efficiently.

The Advanced UPS could be a much, much higher capacity version made with the kind of crazy stuff that you find in the labs Lets say it uses superconductor loops or something crazy like that to hold, I dunno, 100K charge? It may or may not also be able to derive charge from the plutonium cells (or whatever those are called now)

The UPS conversion mod is used to make a tool draw power directly from the UPS rather than any loaded batteries. (We may want to make the UPS wearable so it doesn't clutter your inventory anymore at that point.) The advantage of this is that you can carry several tools that all draw power from a central, high capacity battery pack, rather than carry three or four different battery reloads with you of different sizes. A device like this would deserve the name of Unified Power System.

**Describe alternatives you've considered**  
Removing the UPS completely is something that was tried once, and people were unhappy about it.

**Additional context**  
Admittedly, this is a drastic change in how the UPS works. This is the missing C++ component of the battery overhaul. It's one thing to design it in text like I did above. It's another thing entirely to actually code it, which... Well, I suppose I should do it, but I have no idea how yet.

**Comment by esotericist** [comment]
Just so it's available for discussion, let me link to [the instructions I was given last time this was being investigated](https://github.com/CleverRaven/Cataclysm-DDA/pull/29983#issuecomment-487778539).

Please don't interpret this as me arguing for or against anything; that post is just a summary of what I was told the goal was by Kevin and Mark. I don't particularly have any strong opinions about it.

**Comment by Inglonias** [comment]
Very helpful to see that. Thank you!

**Comment by TechyBen** [comment]
As long as it's a convenience tool, that sounds reasonable. No need to overdo the stats (with the exception of plutonium power, depending if we are going for realism or not etc).

But it would allow you to hold a nice size of charge, and power multiple devices, without swapping batteries (but would add the needed mods to the tools).

Likewise, two of them could be swapped out if it has a charging time required, or just swap out/charge the main battery.

Something like the Plutonium powered version, could charge slowly (automatically once loaded with a rod), but hold a larger/medium charge. That would make it more realistic than the Plutonium just providing an instant massive power source.

**Comment by Inglonias** [comment]
I was thinking of actually redoing the advanced UPS as just an ultra high capacity version, so we can do away with the atomic stuff. It's just a very power-dense black box (I used the term "superconductive loops")

**Comment by TechyBen** [comment]
We already have an atomic coffeemaker. So syphoning such a device to power something should be possible (in game lore), just heavy and slow (as IRL in the likes of the Mars Rovers) to build a useable charge on more power hungry tools.

**Comment by kevingranade** [comment]
As Eso pointed out, the solution I want is for UPS compatible tools to draw
power directly from the UPS when used. Generally these tools would not have
a battery of their own at all, and because of that would be significantly
lighter than battery-powered equivalents.

Shunting charge around manually is both not representative of any real
system and a hassle UI wise.

**Comment by Inglonias** [comment]
> As Eso pointed out, the solution I want is for UPS compatible tools to draw power directly from the UPS when used. 

And that is the plan. But having the ability to discharge batteries into one another is also a nice-to-have.

**Comment by esotericist** [comment]
> > As Eso pointed out, the solution I want is for UPS compatible tools to draw power directly from the UPS when used.
> 
> And that is the plan. But having the ability to discharge batteries into one another is also a nice-to-have. There is currently no way to do that and since you'll need to be able to charge these things anyhow...

While I haven't been doing a lot of coding recently, on my list of Things To Do is a tool for funneling power from one source to another, where you load the tool with the object containing charge, then use the tool on another object that can contain charge. For an analogy in gameflow terms, it'd be a tool for "repairing" a battery with another battery, but with charge instead of durability.

Also on my list of Things To Investigate is a device analogous to the recharging station but for discharging loose power-containing objects into the vehicle.

Both of the above things would almost certainly involve an efficiency cost to some degree.

Either way, mechanisms for moving charge around are definitely desirable, but they don't particularly need to be attached to the UPS mechanism.

**Comment by Inglonias** [comment]
> Either way, mechanisms for moving charge around are definitely desirable, but they don't particularly need to be attached to the UPS mechanism.

The new UPS doesn't use battery magazines right now. Part of the point of the UPS is that you don't need to mess around with dozens of batteries anymore, so why not go all the way with that? Make it so you can dump all your batteries into the UPS and then only carry that, instead of the batteries. 

Right now, I'm hearing a lot of "I don't know if I'm ok with the UPS being able to just transfer charge at will between batteries". I feel like having that ability would enhance it and make sense for a device designed to bridge the gap between all power systems.

It sounds like the decision is between these three options:

1. The UPS takes heavy battery magazines and can also seamlessly power UPS compatible tools. This may be a bit of a hassle to get working properly.
2. The UPS has an integrated battery that can drain from any kind of battery via an iuse, but can also seamlessly power UPS compatible tools. This sounds like the easiest solution to me, but also sort of sidesteps the idea of battery magazines a bit.
3. The UPS has an integrated battery that can drain from and recharge any kind of battery via an iuse, but can also seamlessly power UPS compatible tools. This was my original plan.

**Comment by esotericist** [comment]
I was under the impression a design goal was to make the UPS system not be the ultimate "use it for everything" it was in the "meta" prior to the battery magazine change. 

That there should be tradeoffs between using UPS and not using UPS, rather than simply "have you achieved full UPS yet".

Is this a design goal you object to?

**Comment by TechyBen** [comment]
> As Eso pointed out, the solution I want is for UPS compatible tools to draw power directly from the UPS when used. Generally these tools would not have a battery of their own at all, and because of that would be significantly lighter than battery-powered equivalents. Shunting charge around manually is both not representative of any real system and a hassle UI wise.

Yep. Seems a handy idea. The UPS should not be a "end game uber overpowered" device. It's basically a portable power bench supply with a voltage cut off (so one of those lipo chargers ;) ) as far as I can tell.

Not something impossible to build. And IMO the weight + encumbrance (or volume if in a pack) trade offs are a good balance. Add to that the possibility of power loss in charging, and it would make a portable option, vs a welding rig/base power system (more efficient, but bigger and static).

**Comment by Inglonias** [comment]
> I was under the impression a design goal was to make the UPS system not be the ultimate "use it for everything" it was in the "meta" prior to the battery magazine change.
> That there should be tradeoffs between using UPS and not using UPS, rather than simply "have you achieved full UPS yet".
> Is this a design goal you object to?

Not especially, no. I agree with it, in fact, but it looks like we disagree about (or I don't understand) what makes the UPS overpowered, and what aspects of my design proposal you disagree with.

With the system I have in mind, you can dump batteries into the UPS, but you still need to bring the UPS with you whenever you want to use UPS modded tools. Is that acceptable? If that is acceptable, is it also acceptable that you can dump power out of the UPS into batteries again?

**Comment by esotericist** [comment]
To reiterate: I don't actually have strong opinions on what happens here. I'm providing information where I have it, and asking questions where it seems valuable to ask the question.

I wouldn't be terribly concerned if the UPS stuff was completely removed (and you might recall I tried to do that at one point before getting yelled at a bunch).

That said: what you're describing is still basically making the UPS an end-point to progression where you no longer want battery replacements for any of your individual tools. If you *can* use the UPS for everything without meaningful cost, then it obviously logically follows you *should* use the UPS for everything, and that's where players ended up.

In what fashion is the scenario you're describing different from the previous "battery end-game"?

**Comment by Inglonias** [comment]
> In what fashion is the scenario you're describing different from the previous "battery end-game"?

Previously, you didn't even need to bring the UPS with you because your tools had their own battery capacity. If this proposal were implemented, you would. Admittedly, that's a small change from what things were like before. Storage space is cheap in the mid and late game.

> That said: what you're describing is still basically making the UPS an end-point to progression where you no longer want battery replacements for any of your individual tools. If you can use the UPS for everything without meaningful cost, then it obviously logically follows you should use the UPS for everything, and that's where players ended up.

The thing is, I don't actually see any other way the UPS can be a useful device. Why the heck would it have even been invented if it didn't actually make using power on the go any easier?  I imagine it being a huge battery with wires that you can plug your tools into. A portable wall outlet. If it isn't that, then what is it? What's the point?

To put it simply, I need to know what the UPS should and should not do, because it sounds to me like I don't understand.

**Comment by DemAvalon** [comment]
> I imagine it being a huge battery with wires that you can plug your tools into. A portable wall outlet.

That sounds to me like the very definition of "Universal Power Supply"

**Comment by kevingranade** [comment]
> But having the ability to discharge batteries into one another is also a nice-to-have. 

It's a nice-to-have that doesn't belong in the UPS.

> The new UPS doesn't use battery magazines right now. Part of the point of the UPS is that you don't need to mess around with dozens of batteries anymore, so why not go all the way with that? Make it so you can dump all your batteries into the UPS and then only carry that, instead of the batteries.

That's called "recharging", and it's something you do *to* the UPS, not something it *does*.
The UPS is a kind of battery, not a thing that manipulates batteries.

> Right now, I'm hearing a lot of "I don't know if I'm ok with the UPS being able to just transfer charge at will between batteries".

I wasn't emphatic enough then, I do not want the UPS to be a tool for transferring charge between batteries.

> Let me know which one is most preferable, and I'll go for that one:

This one
>    The UPS takes heavy battery magazines and can also seamlessly power UPS compatible tools. This may be a bit of a hassle to get working properly.

**Comment by I-am-Erk** [comment]
I like the idea of a UPS being a large, powerful battery maybe with a wall outlet on it for powering tools that don't usually have a battery pack.... Using a couple heavy battery magazines to have a huge capacity. Then some currently totally illogical tools, like battery powered jackhammers, could run off ups and actually be wall socket powered.

**Comment by Inglonias** [comment]
I was afraid you'd say you prefer the UPS to take battery magazines. When I said "this will be a hassle" what I meant was "I literally have no idea how to do this"

**Comment by kevingranade** [comment]
Converting tools that expect wall outlet output to be UPS powered would be exactly the kind of thing the UPS would want to do.

> I imagine it being a huge battery with wires that you can plug your tools into. A portable wall outlet. If it isn't that, then what is it? What's the point?

That's exactly what it does, but the tools you plug into it don't have batteries of their own.

**Comment by Inglonias** [comment]
Alrighty, so I've done some testing.

1. It looks like this already works fine for most tools. Specifically, those that have an iuse. The tools it doesn't work for are mainly the lighting tools. Things like flashlights that don't really have an iuse.
2. Just saw #31370 while testing this. I'm starting to think that my proposed changes aren't welcome at the moment, and this needs to be adjusted, so I'm closing this issue and opening another one with a more accurate title.

**Comment by TechyBen** [comment]
> I wasn't emphatic enough then, I do not want the UPS to be a tool for transferring charge between batteries.

Ah that makes sense. So it's not a universal battery input + universal output. It's a single battery (hardwired or swappable depending on final design) and universal output. Thanks. :)
